materials/debugging-R.md

```diff
@@ -5,56 +5,63 @@ title: Debugging
 
-## Setup
-
-**Send `debugging-example-unfixed.R` to students so they can follow along.**
-
-* Start with a new project including `debugging-example-unfixed.R`
-
-## Basic Strategy
-
-* How do we figure out what's wrong with a program
-    * Be a scientist
-	    * Hypothesis about what is wrong
-	    * Make one change that is expected to fix it
-		* Check if it worked
-		* **Do not change something without a reason**
-    * Read the error message
-    * Talk through the code (rubber duck programming)
-	* Observe what the code is doing
-
-
-## Observe
-
-*Run*
-
-* Read the error message and try to fix it
-* Where did code fail? Show traceback
-* Read the code
-* Look at the current state of the environment
-* This gives us a snapshot of what's going on in the code
-* Can't see inside the function
-
-
-## Run line by line
-
-* Run the code line by line checking each step
-* Trick once functions are involved
-
-
-## Print statements
+> Remember to
+>
+> * download [`debugging-example-unfixed.R`]({{ site.baseurl }}/materials/debugging-example-unfixed.R) and [`debugging-example-fixed.R`]({{ site.baseurl }}/materials/debugging-example-fixed.R).
+> * send `debugging-example-unfixed.R` to students so they can follow along.
+> * set up a new `Git` project that includes `debugging-example-unfixed.R`.
+
+## Basic Manual Debugging Strategy
+
+* How do we figure out what's wrong with a program?
+    * Be a scientist.
+        * Hypothesize about what is wrong.
+        * Make one change that is expected to fix error.
+        * Check if change worked/fixed error.
+    * Read the error message.
+    * Talk through the code
+        * Observe what the code is doing
+        * [Rubber duck programming](https://en.wikipedia.org/wiki/Rubber_duck_debugging)
+    * **Do not change something without a reason.**
+
+### Observe
+
+> Run `debugging-example-unfixed.R`.
+
+* Read the error message and try to fix it.
+* Where did code fail? 
+    * Traceback
+    * `options(error = tamper::tamper)` helps with pipes (`%>%`)
+* Read the code.
+* Look at the current state of the environment.
+    * This gives us a snapshot of what's going on in the code
+    * Can't see variables inside the function
+
+### Run line by line
+
+* Run the code line by line checking each step.
+* More difficult once functions are involved.
+    * Isolate functions.
+    * Use `print()`.
+
+### `print()` statements
 * Show us things that we can't see after error
     * Dynamic changes in code
-	* Things inside functions
+    * Things inside functions
+
+```
+print(head(df))
+``` 
 
-*`print(head(df))`, fix by passing `data_size_class`*  
-*Run*  
-*Something seems weird about these results*
-	
+> Fix error by passing `data_size_class` instead of `data`.
+> Run `debugging-example-unfixed.R`.  
+
+* Something seems weird about these results
+    * difficult to tell with manual strategy	
 
 ## Debugger
 
-* Like a print statement but better
+* Advanced error visualization (`print()`) technique
```

Comment:
I like the original version better here. I'm open to a revision but without knowing what was there before I wouldn't understand what this was meant to say. For lecture I think that simple intros like "like X, but better" can be fine.